opendevin/memory/history.py

```diff
@@ -1,54 +1,260 @@
 
 
-class ShortTermHistory:
+class ShortTermHistory(list[Event]):
     """
-    The short term history is the most recent series of events.
-    An agent can send this in the prompt or use it for other purpose.
+    A list of events that represents the short-term memory of the agent.
+
+    This class provides methods to retrieve and filter the events in the history of the running agent from the event stream.
     """
 
+    start_id: int
+    end_id: int
+    _event_stream: EventStream
+    delegates: dict[tuple[int, int], tuple[str, str]]
+    filter_out: ClassVar[tuple[type[Event], ...]] = (
+        NullAction,
+        NullObservation,
+        ChangeAgentStateAction,
+        AgentStateChangedObservation,
+    )
+
     def __init__(self):
+        super().__init__()
+        self.start_id = -1
+        self.end_id = -1
+        self.delegates = {}
+
+    def set_event_stream(self, event_stream: EventStream):
+        self._event_stream = event_stream
+
+    def get_events_as_list(self) -> list[Event]:
         """
-        Initialize the empty list of events
+        Return the history as a list of Event objects.
         """
-        self.events = []
+        return list(self.get_events())
 
-    def add_event(self, event_dict: dict):
+    def get_events(self, reverse: bool = False) -> Iterable[Event]:
         """
-        Adds an event to memory if it is a valid event.
+        Return the events as a stream of Event objects.
+        """
+        # TODO handle AgentRejectAction, if it's not part of a chunk ending with an AgentDelegateObservation
+        # or even if it is, because currently we don't add it to the summary
+
+        # iterate from start_id to end_id, or reverse
+        start_id = self.start_id if self.start_id != -1 else 0
+        end_id = (
+            self.end_id
+            if self.end_id != -1
+            else self._event_stream.get_latest_event_id()
+        )
+
+        for event in self._event_stream.get_events(
+            start_id=start_id,
+            end_id=end_id,
+            reverse=reverse,
+            filter_out_type=self.filter_out,
+        ):
+            # TODO add summaries
+            # and filter out events that were included in a summary
 
-        Parameters:
-        - event_dict (dict): The event that we want to add to memory
+            # filter out the events from a delegate of the current agent
+            if not any(
+                # except for the delegate action and observation themselves, currently
+                # AgentDelegateAction has id = delegate_start
+                # AgentDelegateObservation has id = delegate_end
+                delegate_start < event.id < delegate_end
+                for delegate_start, delegate_end in self.delegates.keys()
+            ):
+                yield event
 
-        Raises:
-        - AgentEventTypeError: If event_dict is not a dict
+    def get_last_action(self, end_id: int = -1) -> Action | None:
```

Comment:
Rename to plural like `get_last_actions` would be clearer, because it _can_ be used for more than one? See `get_last_events` (line 144) for example.

Reply:
it returns `next(...)` action, so only one?

Reply:
Oh, I see. Is the `end_id` used then at all differently?